### ISSUE TITLE
Default to python.analysis.typeCheckingMode: strict

### ISSUE THREAD

**Issue by OlafvdSpek** [microsoft/pylance-release#1743]
Could this default to strict?
Not sure why it defaults to off, it's quite a useful feature if I say so myself.

// Defines the default rule set for type checking.
//  - off: Surfaces diagnostics for invalid syntax, unresolved imports, undefined variables.
//  - basic: All "off" rules + basic type checking rules.
//  - strict: All "off" rules + all type checking rules.
"python.analysis.typeCheckingMode": "off",

**Comment by jakebailey** [comment]
The choice of "off is intentional. The bulk of Python users are not interested in type checking, and the ecosystem itself isn't type-complete for a wide range of libraries (or individual codebases), so we'd be producing misleading diagnostics. This would be like VS Code's JavaScript/TypeScript support changing the default of the `checkJs` setting to true, enabling diagnostics in previously un-checked un-typed code, except without the large ecosystem of typed libraries that TypeScript has amassed over many years.

We are working on ways to better notify users that type checking can be enabled, though, as we know it should be more discoverabe.

**Comment by jakebailey** [comment]
I'll also note that if we were to recommend a level, we'd recommend "basic"; "strict" is very, very strict, and only the most well-structured codebases will pass that mode.

**Comment by OlafvdSpek** [comment]
I thought unavailability of typing data wouldn't cause diagnostics.. 
In that case, a mode where diagnostics are shown only when typing info is available might make sense.

**Comment by erictraut** [comment]
Diagnostic rules are fine-grained, so you can choose which ones you want to enable for your project. Here's a [full list of the diagnostic rules](https://github.com/microsoft/pyright/blob/main/docs/configuration.md). The `typeCheckingMode` setting provides useful collections of defaults for each of these diagnostic checks. Refer to [this table](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#diagnostic-rule-defaults) for the mapping. You can start with these defaults and override them (e.g. enable additional diagnostic rules) if you'd like.

**Comment by jakebailey** [comment]
> I thought unavailability of typing data wouldn't cause diagnostics..

That'd be closest to `"useLibraryCodeForTypes": false` without "strict" (as that disallows unknown types), which then treats untyped libraries as `Any`, yes, but that doesn't change the fact that users open up their own codebases and will see type errors that weren't there before.

> Having Pylance and not having any warning for this seems just weird.

That code is pretty simple and clear that it would fail, but real-world code get a lot more tricky. Write the same sort of code in JS, and you'll also see that VS Code won't emit a diagnostic for that either, unless you opt-in to `checkJs` or switch to TypeScript.

**Comment by erictraut** [comment]
> Having Pylance and not having any warning for this seems just weird.

As @jakebailey said, the vast majority of Pylance users don't want to see type warnings or errors. They're interested in completion suggestions, contextual doc strings, and other development efficiency features.

Users who are interested in type checking warnings and errors are typically more sophisticated Python users, so it's probably reasonable to ask them to enable the type checking functionality. We're exploring ways that we can make this functionality more discoverable.

**Comment by OlafvdSpek** [comment]
> but that doesn't change the fact that users open up their own codebases and will see type errors that weren't there before.

You mean the diagnostics weren't there? ;)
If the errors aren't false positives, I'm not sure why you wouldn't want to know about them.

 > > Having Pylance and not having any warning for this seems just weird.
> 
> That code is pretty simple and clear that it would fail, but real-world code get a lot more tricky. 

Sure. If a rule can't be enabled without false positives I understand not enabling it by default.

**Comment by jakebailey** [comment]
The "off" mode already enables the small subset of diagnostics we believe don't have false positives, that is, import resolution issues and undefined variable usage. We previously even warned on potentially unbound variables, but had to disable it as people complained that their "valid" (it runs) code was generating diagnostics. If we can't even get unbound variable detection enabled without feedback saying to disable it, I don't see us being able to enable anything stronger.

**Comment by shoffmeister** [comment]
I installed `mypy` simply because I was unaware of Pylance _having_ type-checking and doing `"python.analysis.typeCheckingMode": "off"` by default (I did not read https://marketplace.visualstudio.com/items?itemName=ms-python.vscode-pylance). 

For me Pylance was the "we switch the extension from Python to Pylance" dance, and, honestly, I had no cycles to discover what Pylance actually _offers_. It did not regress, and that was all fine and well then.

So, I do agree that adding better discoverability could help, perhaps as part of the Python walk-through? Or upon sensing that a user has type information in a Python file, offering a tour through the typing support?
